cilium/CFP-25694-move-cilium-cli-to-cilium-repo.md

```diff
@@ -0,0 +1,145 @@
+# CFP-25694: Move cilium/cilium-cli code into cilium/cilium repository
+
+**Code Name: Mega Ketchup**
+
+**SIG: [sig-cilium-cli]**
+
+**Begin Design Discussion:** 2023-05-25
+
+**Cilium Release:** 1.15
+
+**Authors:** Michi Mutsuzaki `<michi@isovalent.com>`
+
+## Summary
+
+Move [cilium/cilium-cli] code into [cilium/cilium] repository.
+
+## Motivation
+
+The main motivation for moving [cilium/cilium-cli] code into [cilium/cilium]
+repository is to minimize the feature development overhead for adding new tests
+to `cilium connectivity test`.
+
+Currently, adding a new test to `cilium connectivity test` requires the following
+steps:
+
+- Open a pull request against [cilium/cilium-cli] repo to add a new test, and
+  wait for the CI build to finish.
+- Modify GitHub workflows files to pick up the CI version of `cilium-cli` in
+  your pull request against [cilium/cilium] repo.
+- Once the new tests passes, undo the change in the workflow files.
+- Get these pull requests reveiewed and merged.
+
+Even after all these steps, the new test does not become a part of the regular
+CI until the [sig-cilium-cli] team releases a new version of `cilium-cli` that
+includes the test, and Renovate updates GitHub wokflow files to pick up the new
+`cilium-cli` version. This has been a major source of frustration for some Cilium
+contributors.
```

Comment:
My recollection of the origins of Cilium CLI in general was that it's intended to focus on the user experience in order to simplify installing, observing, maintaining live Cilium clusters. Tools like `cilium connectivity test` were conceived as a sanity check for user environments so we can quickly see whether a prod environment is healthy or not.

Over time, given that this framework provided mechanisms to develop new tests without the baggage of ginkgo and jenkins, I think it became an attractive place for Cilium developers to expand testing. Now we're at several minutes to assess the cluster sanity, and it's doing a lot of advanced feature testing. We're talking about reconfiguring the target environment to perform some of this testing, and starting to hide such "extra test" options behind special flags to avoid impacting prod environments.

I have a bit of concern that over time we're losing sight of the original UX-focused goals of Cilium-CLI and turning it into the Cilium developer CI tool instead. On the way on that path, we'll simplify some Cilium developer workflows (ideal) but also potentially raise the barrier to entry for cilium-cli contribution (not ideal).

Should the tests for advanced Cilium functionality exist in cilium/cilium or cilium/cilium-cli? Well, probably cilium/cilium. Does that mean that all of cilium-cli should then go into cilium/cilium? I'm not so sure. I'd hope that most testing can be done in unit or module / cell level tests directly in the tree, rather than being e2e. Even if we do need e2e tests for some of the new functionality, is the user-facing tool the right place to implement a lot of that advanced testing?

Reply:
> Could we model the cilium/cilium CI like this instead, and autopublish CLI images for every PR or every main commit... then provide a way for cilium/cilium PRs to point to those CLI dev images?

we already do that, sort of. you can specify a commit SHA of cilium-cli here: https://github.com/cilium/cilium/blob/6d4b2f7092d634f67c6192c85143051c49846245/.github/workflows/conformance-aks.yaml#L74. you just need to edit these workflow files. i guess some people find this step more annoying than others.

>  I'd hope that most testing can be done in unit or module / cell level tests directly in the cilium/cilium tree, rather than being e2e. Even if we do need e2e tests for some of the new functionality, is the user-facing tool the right place to implement a lot of that advanced testing?

i think this is the key question. `cilium connectivity test` unintentionally became "the" place to add e2e tests. maybe that's where things started going wrong 🔥 cc @brb.

Reply:
@gandro had another idea: use cilium-cli's `Hook` interface https://github.com/cilium/cilium-cli/blob/main/internal/cli/cmd/hooks.go. this might make more sense if we want to keep cilium-cli relatively lean, and at the same time let cilium developers use the same `cilium connectivity test` framework to add advanced e2e tests.

Reply:
> then provide a way for cilium/cilium PRs to point to those CLI dev images?

I added this - https://github.com/cilium/cilium/pull/25228. However, bunch of workflows `# pull_request: {}` need to be uncommented though to make changes effective.

Reply:
> @gandro had another idea: use cilium-cli's `Hook` interface

I like also this idea :bulb: keep the connectivity test in cilium-cli relatively contained and have:
* a cilium-cli in OSS cilium that extends cilium/cilium-cli, like we are currently doing with CEE, with the full connectivity test suite
* a cilium-cli in CEE cilium that extends the OSS one with the enterprise specific testing bits

in this way each repo would own its tests (and CEE can still run the full OSS suite).

One problem I see with that approach is that it still doesn't solve the current cilium-cli CI issue, as it's possible to miss breakages until the CLI changes are tested in the cilium/cilium workflows.

We could go either with the renovate approach, which while not optimal, would still improve the current state of things, as at least once a week cilium/cilium would be synced with the most recent CLI revision, or we could add an additional workflow in the CLI that pulls latest cilium/cilium main, bumps the CLI dependency to the CLI PR revision and run the full e2e suite (even though that doesn't work with CEE I guess)

Reply:
> at least once a week cilium/cilium would be synced with the most recent CLI revision

Unfortunately, everything which gets tested after a merge usually involves a few folks spending their cycles instead of a single PR owner.